tss2-dlopen/tss2-dlopen-esys.c

```diff
@@ -0,0 +1,2199 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*******************************************************************************
+ * Copyright 2021, Fraunhofer SIT
+ * All rights reserved. *******************************************************************************/
+
+/**
+ * The purpose of this file is to copy it into your project and
+ * include it during compilation if you don't want to link against
+ * libtss2-esys at compile time.
+ * It will attempt to load libtss2-esys.so during runtime.
+ * It will either work similarly to directly linking to libtss2-esys.so
+ * at compile-time or return a NOT_IMPLEMENTED error.
+ *
+ * For new versions of this file, please check:
+ * http://github.com/tpm2-software/tpm2-tss/tss2-dlopen
+*/
+
+#include <dlfcn.h>
+#include <stdio.h>
+#include <tss2/tss2_esys.h>
+
+#define str(s) xstr(s)
+#define xstr(s) #s
+
+#ifdef ENABLE_WARN
+#define WARN(str, ...) fprintf(stderr, "WARNING: " str "\n", ## __VA_ARGS__);
+#else /* ENABLE_WARN */
+#define WARN(...) ;
```

Comment:
change do: `do {} while(0)` or so?

That way people can put a semicolon at the end and will not get the right number of expressions and semicolons, so that you can then do this safely:

```c
if (x)
        WARN(…);
```

And you'll get exactly one ";" generated and no uncertainty what is done as part of the if branch and what not